### PR TITLE
Remove tech debt related to image disposal and layer GC

### PR DESCRIPTION
**Pull request by dnfield**
WIP - do not land until after https://github.com/flutter/flutter/pull/84740

- [x] don't forget to add some test(s) to cover images getting cleaned up when picture/enginelayer is disposed.

Fixes https://github.com/flutter/flutter/issues/84116

### CODE REVIEW

File: lib/ui/painting/image_dispose_unittests.cc [hunk 1]
```diff
@@ -30,23 +30,11 @@ class ImageDisposeTest : public ShellTest {
...
 };
 
-TEST_F(ImageDisposeTest,
-#if defined(OS_FUCHSIA)
-       DISABLED_ImageReleasedAfterFrame
-#else
-       ImageReleasedAfterFrame
-#endif  // defined(OS_FUCHSIA)
-) {
+TEST_F(ImageDisposeTest, ImageReleasedAfterFrameAndDisposePictureAndLayer) {
```
Comment (dnfield) [inline lib/ui/painting/image_dispose_unittests.cc:37]: Reviewers take note: this re-enables a test that was flaking on Fuchsia.

However, this test no longer relies on GC behavior and should be fully deterministic now.

Comment (dnfield) [reply lib/ui/painting/image_dispose_unittests.cc:37]: It also no longer relies on decoding a massive image, and is much faster now too.

File: lib/ui/painting/engine_layer.cc [hunk 2]
```diff
@@ -25,16 +25,6 @@ EngineLayer::EngineLayer(std::shared_ptr<flutter::ContainerLayer> layer)
...
 EngineLayer::~EngineLayer() = default;
 
-size_t EngineLayer::GetAllocationSize() const {
-  // TODO(dnfield): Remove this when scene disposal changes land in the
-  // framework. https://github.com/flutter/flutter/issues/81514
-
-  // Provide an approximation of the total memory impact of this object to the
-  // Dart GC.  The ContainerLayer may hold references to a tree of other layers,
-  // which in turn may contain Skia objects.
-  return 3000;
```
Comment (zanderso) [inline lib/ui/painting/engine_layer.cc:35]: Is the shallow size of the C++ object still tracked as an external allocation for Dart GC?

File: lib/ui/painting/image_dispose_unittests.cc (hunk 1 above)
Comment (zanderso) [reply lib/ui/painting/image_dispose_unittests.cc:37]: https://github.com/flutter/flutter/issues/84116 can be closed when this lands.

**Review by zanderso — COMMENTED** [review]
Dart may wish to deprecate the `Dart_HintFreed()` API as well after this lands. /cc @rmacnak-google

File: lib/ui/painting/engine_layer.cc (hunk 2 above)
Comment (dnfield) [reply lib/ui/painting/engine_layer.cc:35]: That's handled automatically here: https://github.com/flutter/engine/blob/master/third_party/tonic/dart_wrappable.h#L89

**Comment by dnfield** [comment]
The change in the framework landed. This is ready for review.

**Review by bdero — APPROVED** [review]
lgtm! It's nice to see this go away. Hopefully my understanding is correct here with respect to the picture lifecycle work.

File: runtime/runtime_controller.cc [hunk 3]
```diff
@@ -195,17 +195,14 @@ bool RuntimeController::ReportTimings(std::vector<int64_t> timings) {
...
   std::shared_ptr<DartIsolate> root_isolate = root_isolate_.lock();
   if (!root_isolate) {
     return false;
   }
 
   tonic::DartState::Scope scope(root_isolate);
 
-  // Dart will use the freed hint at the next idle notification. Make sure to
-  // Update it with our latest value before calling NotifyIdle.
-  Dart_HintFreed(freed_hint);
```
[the remark this answers is not in the document]
Comment (dnfield) [reply runtime/runtime_controller.cc:208]: Correct. Running the GC for this is expensive and no longer needed

**Pull request merged**